snapdtool/tool_linux.go

```diff
@@ -219,6 +219,31 @@ func ExecInSnapdOrCoreSnap() {
 	}
 
 	logger.Debugf("restarting into %q", full)
+
+	// We want to make "ps", "top" and other tools show a
+	// command-line that is not misleading.
+	originalDir, originalBase := filepath.Split(os.Args[0])
+	// In the case of symlink, typically /snap/bin/myapp ->
+	// /usr/bin/snap, we want to keep the original path as the
+	// user will want to know what they originally intended to
+	// execute. More importantly, we will read os.Args[0] to
+	// decide what application to effectively run. So we must do
+	// nothing in that case.
+	if originalBase == filepath.Base(full) {
+		// Otherwise...
+		// If we did not have any / in the path, it was
+		// executed from PATH. So we do not have to change it.
+		// For instance "snap list" should stay "snap list".
+		if originalDir != "" {
+			// In the other case, we probably executed
+			// from a fork or from a service.  For example
+			// /usr/lib/snapd/snapd from snapd.service.
+			// In this case keeping the original path
+			// would be misleading. So let's change it.
+			os.Args[0] = full
```

Comment:
this branch isn't covered by unit tests, given how important the reexec part is please add a test or 2 covering the new code

Reply:
Done.